tfhe/src/core_crypto/entities/modulus_switched_lwe_ciphertext.rs

```diff
@@ -0,0 +1,286 @@
+            };
+
+        assert!(
+            log_modulus <= uncompressed_ciphertext_modulus_log,
+            "The log_modulus (={log_modulus}) for modulus switch compression must be smaller than the uncompressed ciphertext_modulus_log (={uncompressed_ciphertext_modulus_log})",
+        );
+
+        let lwe_size = ct.lwe_size().0;
+
+        let number_bits_to_pack = lwe_size * log_modulus;
+
+        let len = number_bits_to_pack.div_ceil(Scalar::BITS);
+
+        let slice = ct.as_ref();
+
+        //                          log_modulus
+        //                           |-------|
+        //
+        // slice        :    |   k   |  k+1  |  k+2  |
+        // packed_coeffs:  i-1   |       i       |     i+1
+        //
+        //                       |---------------|
+        //                         Scalar::BITS
+        //
+        //                   |---|
+        //                start_shift
+        //
+        //                       |---|
+        //                       shift1
+        //                (1st loop iteration)
+        //
+        //                       |-----------|
+        //                           shift2
+        //                    (2nd loop iteration)
+        //
+        // packed_coeffs[i] =
+        //                    slice[k] >> start_shift
+        //                    | slice[k+1] << shift1
+        //                    | slice[k+2] << shift2
+        //
+        // with << corresponding to shifting bits on the right
+        // and  >> corresponding to shifting bits on the left
+        let packed_coeffs = (0..len)
+            .map(|i| {
+                let mut j = Scalar::BITS * i / log_modulus;
+
+                let start_shift = i * Scalar::BITS - j * log_modulus;
+
+                let mut value = switch_modulus(slice[j]) >> start_shift;
+                j += 1;
+
+                while j * log_modulus < ((i + 1) * Scalar::BITS) && j < lwe_size {
+                    let shift = j * log_modulus - i * Scalar::BITS;
+
+                    value |= switch_modulus(slice[j]) << shift;
+
+                    j += 1;
+                }
+                value
+            })
+            .collect();
+
+        let log_modulus = CiphertextModulusLog(log_modulus);
+
+        Self {
+            packed_coeffs,
+            lwe_dimension: ct.lwe_size().to_lwe_dimension(),
+            log_modulus,
+            uncompressed_ciphertext_modulus,
+        }
+    }
+
+    /// Converts back a compressed ciphertext to its initial modulus
+    /// The noise added during the compression says int hte output
+    /// The output must got through a PBS to reduce he noise
+    pub fn extract(&self) -> LweCiphertextOwned<Scalar> {
+        let log_modulus = self.log_modulus.0;
+
+        // log_modulus lowest bits set to 1
+        let mask = Scalar::MAX >> (Scalar::BITS - log_modulus);
+
+        let container = (0..(self.lwe_dimension.0 + 1))
```

Comment:
self.lwe_dimension.to_lwe_size().0 ?

it's not critical, but I would prefer avoiding manual operations on dimensions in the code base